### PR TITLE
Loadtest users are like demo users

### PR DESCRIPTION
**Pull request by kaapstorm**
## Product Description

Makes the "Loadtest users" feature Generally Available for Pro Plan subscriptions and higher.

This change adds some guard rails:

* Loadtest users are unable to sync payloads over 500,000 cases
* The "loadtest factor" field is moved from the mobile worker's Basic Info form, to under the "Actions" tab, where it is less likely to cause confusion for users just trying to create a normal mobile worker.
* Loadtest users are like demo users, in that their form submissions are ignored and they are unable to sync from a mobile device.

## Technical Summary

Extends PR https://github.com/dimagi/commcare-hq/pull/31314

The changes in this PR are based on QA feedback, and the discussion that it prompted.

It is already difficult for loadtest users to break real cases. (They have to set the real case as an extension case of a loadtest case (which requires an app that supports that.) This would result in an index that points to a host case that does not exist.)

This PR proposes using the "demo users" functionality to prevent mobile workers from making any changes to cases at all.

Probably easiest to review by commit. :blowfish: Commits specific to this PR start at [Revert "reset_loadtest_factor()"](392a5e87691216f2689e3e82652fe29786b74e08).

See the original PR https://github.com/dimagi/commcare-hq/pull/31095 for more context.

## Feature Flag

Was "Loadtest Users", now Pro Plan or higher.

### Safety story

The total number of loadtest cases is capped.

Loadtest users are unable to make changes to cases.

More details on the instructions and guidelines that we provide can be found in [Confluence](https://confluence.dimagi.com/display/saas/Loadtest+Users).

And there is more context in [this Google doc](https://docs.google.com/document/d/15nNnRAh8vx4v5ds9VF9cJHmlLE-NoFi1P0THKI86UCk/edit#) on the process we followed to determine this behavior.

### QA Plan

This feature has passed through QA.

### Rollback instructions

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

### CODE REVIEW

File: corehq/apps/hqcase/tests/test_loadtest_users.py [hunk 1]
```diff
@@ -22,23 +31,25 @@ class LoadtestUserTest(TestCase, DomainSubscriptionMixin):
     @classmethod
     def setUpClass(cls):
         super(LoadtestUserTest, cls).setUpClass()
-        cls.domain = create_domain(DOMAIN)
```
Comment (proteusvacuum) [inline corehq/apps/hqcase/tests/test_loadtest_users.py]: what is the rationale behind this change? If something else isn't cleaning up after itself, can we change that instead?

File: corehq/ex-submodules/casexml/apps/phone/data_providers/case/load_testing.py [hunk 2]
```diff
@@ -1,8 +1,16 @@
+from __future__ import annotations
```
Comment (proteusvacuum) [inline corehq/ex-submodules/casexml/apps/phone/data_providers/case/load_testing.py:1]: I'm not crazy about this, as I don't believe we have a tech-debt process in place here to remove these when we move to 3.10, or whenever this is slated to become the default.

File: corehq/ex-submodules/casexml/apps/phone/restore_caching.py [hunk 3]
```diff
@@ -35,6 +38,9 @@ def invalidate(self):
 
 @quickcache(['domain', 'user_id'], timeout=24 * 60 * 60)
 def get_loadtest_factor_for_user(domain, user_id):
+    # This function is only used for calculating the restore cache key.
+    # It does not determine how many cases a loadtest user will sync.
```
Comment (proteusvacuum) [inline corehq/ex-submodules/casexml/apps/phone/restore_caching.py:42]: Do you think changing the function name would make this clearer?

File: corehq/apps/users/forms.py [hunk 4]
```diff
@@ -468,6 +465,51 @@ def update_user(self, **kwargs):
...
 
 
+class CommCareUserActionForm(BaseUpdateUserForm):
+
+    loadtest_factor = forms.IntegerField(
+        required=False,
+        min_value=1,
+        max_value=LOADTEST_HARD_LIMIT,
+        help_text=gettext_lazy(
+            "Multiply this user's case load by a number for load testing on "
```
Comment (proteusvacuum) [inline corehq/apps/users/forms.py:475]: Maybe "by _this_ number"?

File: corehq/apps/users/tasks.py [hunk 5]
```diff
@@ -346,3 +346,29 @@ def process_reporting_metadata_staging():
     run_again = run_periodic_task_again(process_reporting_metadata_staging_schedule, start, duration)
     if run_again and UserReportingMetadataStaging.objects.exists():
         process_reporting_metadata_staging.delay()
+
+
+@task(queue='background_queue', acks_late=True)
+def loadtest_user_is_demo_user(commcare_user_id):
```
Comment (proteusvacuum) [inline corehq/apps/users/tasks.py:352]: would be nice to add a test for this

**Review by proteusvacuum — CHANGES_REQUESTED** [review]
Looks great! 
GTM from me if you add a test for the demo user functionality.

File: corehq/ex-submodules/casexml/apps/phone/data_providers/case/load_testing.py [hunk 6]
```diff
@@ -19,19 +27,24 @@ def _map_id(id, count):
     return CaseSyncUpdate(case, update.sync_token, required_updates=update.required_updates)
 
 
-def get_xml_for_response(update, restore_state):
+def get_xml_for_response(
+    update: CaseSyncUpdate,
+    restore_state: RestoreState,
+    total_cases: int,
+) -> list[str]:
```
Comment (millerdev) [inline corehq/ex-submodules/casexml/apps/phone/data_providers/case/load_testing.py:34]: Can you explain why these annotations are being added here?

Comment (kaapstorm) [reply corehq/ex-submodules/casexml/apps/phone/data_providers/case/load_testing.py:34]: Those annotations were actually added [in February](https://github.com/dimagi/commcare-hq/pull/31095) (that's how old the original PR is!) prior to our discussions about type hints. They come from before I dropped the type hints from [`livequery.compile_response()`](https://github.com/dimagi/commcare-hq/pull/31095/commits/5a0e99c4fc4bafc2921cb1eff45461025365c9e1). (This work partially motivated those discussions, in the hope that we could get some clarity on how best to use type hints in HQ.)

To answer your question, at the time, I used them so that I had some visibility into exactly what an "update" is, and left them so that the next person might also benefit.

I'll drop them.

But, as you know, I don't think that's a great option. In fact, I don't think _any_ of the options are _great_ options:

1. Leave this as is. Not verified by GitHub Actions. At least verified by IDEs. Helps the next dev unfamiliar with this area of the codebase.

2. Replace this with a docstring:
```python
def get_xml_for_response(update, restore_state, total_cases):
    """
    :param update: An instance of CaseSyncUpdate
    :param restore_state: An instance of RestoreState
    :param total_cases: An integer
    :returns: A list (not a string) of case XML (strings)
    """
```
That docstring is clumsy. It isn't verified by anything, not even an IDE. Not an improvement (imho).

3. Drop entirely:
```python
def get_xml_for_response(update, restore_state, total_cases):
```
The least helpful option. But it is nice and neat. And everyone is OK with this option. So, not perfect, but it'll do.

4. Just leave the useful bits:
```python
def get_xml_for_response(update: CaseSyncUpdate, restore_state, total_cases) -> list[str]:
```
An unholy violation of everything wholesome and beautiful. But useful. Allegedly "practicality beats purity", but how far does that stretch?

Comment (proteusvacuum) [reply corehq/ex-submodules/casexml/apps/phone/data_providers/case/load_testing.py:34]: There's another option, which is self documenting, carries through to the rest of the code, and is not controversial:

```py
def get_xml_for_response(case_sync_update, ...):
```

File: corehq/apps/users/tasks.py [hunk 7]
```diff
@@ -346,3 +346,29 @@ def process_reporting_metadata_staging():
...
+    from corehq.apps.users.models import CommCareUser
+
+    user = CommCareUser.get_by_user_id(commcare_user_id)
+    if user.loadtest_factor and user.loadtest_factor > 1:
+        if not user.is_demo_user:
+            user.is_loadtest_user = True
+            turn_on_demo_mode(user, user.domain)
+    else:
+        if user.is_loadtest_user:
+            user.is_loadtest_user = False
```
Comment (proteusvacuum) [inline corehq/apps/users/tasks.py:373]: This is getting persisted because of the `save` call in `turn_off_demo_mode`, which I think is OK, but was a bit of a surprise to me.

File: corehq/apps/users/tasks.py (hunk 5 above)
Comment (proteusvacuum) [inline corehq/apps/users/tasks.py:352]: With the test, I think this is a hard-to-parse name for this function and it has unexpected behaviour when used in context - I would have thought this was going to return a boolean (answering the question - "is this loadtest user a demo user?"). 
Maybe something like `apply_correct_demo_mode_to_loadtestuser` or something?

**Review by proteusvacuum — APPROVED** [review]

**Comment by kaapstorm** [comment]
Thanks for the great feedback.
